server.js

```diff
@@ -2,22 +2,137 @@ const express = require("express");
 });
+//////////////////////Create a new message//////////////////////////// 
+ app.post('/messages/', (req, res) =>{
+  let newId = messages[messages.length -1].id + 1
+  let newMessage = {
+        id: newId,
+        from: req.body.from,
+        text: req.body.text,
+        created_at :new Date().toISOString()
+      };
+ 
+  if((newMessage.from !== undefined) && (newMessage.from.length > 0) && (newMessage.text !== undefined)&& (newMessage.from.length > 0) ){
+  messages.push(newMessage)
+  res.json(messages)
+    }
+    else{
+      res.status(400).send("Please post a validate message")
+    }
+    })
+////////////////////////Read all messages//////////////////////////
+app.get('/messages', (req, res) => {
+  res.json(messages)
+})
+/////////////////////////Read one message specified by an ID///////////////////////////
+app.get('/messages/:id', (req, res) => {
+  const mesId = Number(req.params.id);
+  const result = messages.find(ms => ms.id === mesId)
+  if (result !== undefined) {
+    res.json(result)
+  }
+  else {
+    res.status(404).send("sorry not found")
+  }
+})
+///////////////////Delete a message, by ID/////////////////////////////////////////////////
+app.delete('/messages/:id', (req, res) => {
+  const delId = Number(req.params.id);
+  const index = messages.findIndex(ms => ms.id === delId)
+  if (index !== undefined) {
+    messages.splice(index, 1)
+    res.json("DELETED!!!")
+  }
+  else {
+    res.status(404).send("sorry not found")
+  }
+})
+////////////////////level 2 /////////////////////////////////
+app.post('/messages/', (req, res) =>{
+  let newId = messages[messages.length -1].id + 1
+  let newMessage = {
+        id: newId,
+        from: req.body.from,
+        text: req.body.text,
+        created_at :new Date().toISOString()
+      };
+ 
+  if((newMessage.from !== undefined) && (newMessage.from.length > 0) && (newMessage.text !== undefined)&& (newMessage.from.length > 0) ){
+  messages.push(newMessage)
+  res.json(messages)
+    }
+    else{
+      res.status(400).send("Please post a validate message")
+    }
+    })
+///////////////////level 3 /////////////////////////////////
+app.get("/messages5/search", function(req, res) {
+  let text = req.query.text
+  // console.log(req.query);
+   if (text !=undefined){
+ const result = messages.filter(word => (word.text.toLowerCase().includes(text.toLowerCase())))
+    res.send(result)
+    // console.log(result);
+  }
+  else{
+    res.send([])
+  }
+});
+////////////updated /////////////////////////////
+app.put('/messages/:id',  (req, res) => {
+  const postId = req.params.id
+  let updatedMessage = {
+    from: req.body.from,
+    text: req.body.text,
+    updated_at :new Date().toISOString()
+  }
+  const checkedMessage = messages.find(u => u.id === parseInt(postId))
+ if(checkedMessage !== undefined){
+  if((updatedMessage.from !== undefined)&&(updatedMessage.from.length > 0) && (updatedMessage.text !== undefined)&&(updatedMessage.text.length > 0) ){
```

Comment:
validation could be factored out to a function and then reused here and in the post message endpoint